### ISSUE TITLE
React to Tools DCR DCR: Change secretKey to something else

### ISSUE THREAD

**Issue by sgellock** [Microsoft/BotFramework-Emulator#866]
DCR: Change secretKey to something else
Microsoft/BotBuilder-Tools#476

**Comment by tonyanziano** [comment]
The work is done, just waiting for the version bump on the `msbot` side.

**Comment by tonyanziano** [comment]
Update:

I've integrated the newest version of `botframework-config` into the emulator and there seems to be an issue inside of the library when calling the node `crypto` module.

When I call `botConfig.validateSecret()`, it will fail inside of the call to `encrypt.encryptString(uuid(), secret)` on this line (https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botframework-config/src/encrypt.ts#L36) with the error:

```
"Invalid key length"
"Error: Invalid key length
    at new Cipheriv (crypto.js:217:16)
    at Object.Cipheriv (crypto.js:215:12)
    at Object.encryptString (C:\Users\toanzian\Desktop\GitRepos\EmuGH\node_modules\botframework-config\src\encrypt.ts:36:42)
    at BotConfiguration.validateSecret (C:\Users\toanzian\Desktop\GitRepos\EmuGH\node_modules\botframework-config\src\botConfiguration.ts:268:40)
    at toSavableBot (C:\Users\toanzian\Desktop\GitRepos\EmuGH\packages\app\main\src\botHelpers.ts:130:12)
    at Object.saveBot (C:\Users\toanzian\Desktop\GitRepos\EmuGH\packages\app\main\src\botHelpers.ts:167:22)
    at <anonymous>"
```

I've been trying to look for a solution online, but haven't found a straightforward fix yet. I'm wondering if it is due to a mismatched node version between the CLI tools and the node that comes with the Electron version the emulator runs on. There was a post I came across that said in a past update, node switched the default encoding for the crypto library from binary to utf-8, so I'm wondering if that has anything to do with it.

I'm assuming the secret encryption is working fine on the CLI side.

**Comment by sgellock** [comment]
@tomlm any ideas?



On Sep 17, 2018, at 6:15 PM, Tony Anziano <notifications@github.com<mailto:notifications@github.com>> wrote:


Update:

I've integrated the newest version of botframework-config into the emulator and there seems to be an issue inside of the library when calling the node crypto module.

When I call botConfig.validateSecret(), it will fail inside of the call to encrypt.encryptString(uuid(), secret) on this line (https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botframework-config/src/encrypt.ts#L36<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FMicrosoft%2Fbotbuilder-js%2Fblob%2Fmaster%2Flibraries%2Fbotframework-config%2Fsrc%2Fencrypt.ts%23L36&data=02%7C01%7Cscott.gellock%40microsoft.com%7C4352bf487aeb4f77047b08d61d042a4d%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636728301070533966&sdata=%2FC6Wmz1JI0kvrPtpRFRA9dfMQpw8qO8Mf%2B4Bhy6eoks%3D&reserved=0>) with the error:

"Invalid key length"
"Error: Invalid key length
    at new Cipheriv (crypto.js:217:16)
    at Object.Cipheriv (crypto.js:215:12)
    at Object.encryptString (C:\Users\toanzian\Desktop\GitRepos\EmuGH\node_modules\botframework-config\src\encrypt.ts:36:42)
    at BotConfiguration.validateSecret (C:\Users\toanzian\Desktop\GitRepos\EmuGH\node_modules\botframework-config\src\botConfiguration.ts:268:40)
    at toSavableBot (C:\Users\toanzian\Desktop\GitRepos\EmuGH\packages\app\main\src\botHelpers.ts:130:12)
    at Object.saveBot (C:\Users\toanzian\Desktop\GitRepos\EmuGH\packages\app\main\src\botHelpers.ts:167:22)
    at <anonymous>"


I've been trying to look for a solution online, but haven't found a straightforward fix yet. I'm wondering if it is due to a mismatched node version between the CLI tools and the node that comes with the Electron version the emulator runs on. There was a post I came across that said in a past update, node switched the default encoding for the crypto library from binary to utf-8, so I'm wondering if that has anything to do with it.

I'm assuming the secret encryption is working fine on the CLI side.

**Comment by sgellock** [comment]
This is on the node side?

**Comment by tonyanziano** [comment]
Yeah it's on the node side.

So I modified the botframework-config library directly to test some possible solutions and I was able to get it to encrypt properly if I used a 32 byte (256 bit) key generated using

crypto.randomBytes(32)

which is the required key length for the aes-256 algorithm being used for encryption.

So I'm not sure how that problem is being avoided on the cli side, but I'm not sure if we should enforce that on the emulator side or check for it in the library.

It's strange to me though that we are running into it now because encryption with this library used to work for the emulator.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by sgellock** [comment]
Are you working with an old bot file?  We had a different key before and there is logic which catches the exception and uses the legacy decryption.

Is it possible you are catching the exception before the fallback logic can process it?
-Tom

**Comment by tonyanziano** [comment]
No I'm using a brand new bot file being created via the new bot flow in the emulator. So there is no secretKey property.

I'm stepping into calls made while calling config.validateSecret() and it throws the error right when it tries to create the cipher for the first time unless I pass it a 32 byte secret.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by tonyanziano** [comment]
Addressed in #898
